lib/galaxy/webapps/galaxy/services/history_contents.py

```diff
@@ -1602,7 +1612,13 @@ def _purge(self, item: HistoryItemModel, trans: ProvidesHistoryContext):
 
     def _change_datatype(self, item: HistoryItemModel, params: ChangeDatatypeOperationParams):
         if isinstance(item, HistoryDatasetAssociation):
-            item.change_datatype(params.datatype)
+            self.hda_manager.ensure_can_change_datatype(item)
+            self.hda_manager.ensure_can_set_metadata(item)
+            item.dataset.state = item.dataset.states.SETTING_METADATA
```

Comment:
You need to flush when you make a change.

Reply:
There is a `global` flush at the end of the request here:

https://github.com/davelopez/galaxy/blob/09543032c7b91bc4ce6b3b488974d8cf3df114fc/lib/galaxy/webapps/galaxy/services/history_contents.py#L710

Should I flush for every item instead?

Reply:
I added the flush here too, just in case, I can revert if the global request flush is enough.